Performance improvements for benchmarks

GabrielMajeri opened the pull request:
This pull request contains improvements to the performance of the example crates in this repository.

## fannkuch_redux

- I noticed a small performance reduction because LLVM was emitting memory accesses to the `self.odd` field, insted of storing it in a register.
  Not sure why this happens, LLVM should know we borrow `self` mutably & have exclusive access to `self` and be able to cache its value locally.

- Overall, about 1.5% better performance.

## stencil

See issue #95 for the full discussion.

- Now we're using `mul_adde` (FMA) where possible.
- Improved performance of `Data::reinit` by using unsafe memory accesses. Not part of the benchmark, but it was slow and showing up a lot in the profiler.
- The parallel code is slower than the SIMD code now:
```
simd: 175 ms
simd+par: 279 ms
ispc: 176 ms
ispc+tasks: 373 ms
```
  I guess that's... good?

### Failing tests
- I've updated the testing code to use approximative float comparisons, with an error no larger than 10^-4.

## aobench

- Added `mul_adde` to most places, matching ISPC's assembly.
- Inlined some functions which weren't getting inlined before.
- Overall, when using the system math library (instead of comparing against ISPC's built-in one) Rust performance is better than with ISPC.

gnzlbg commented:
I think this can be updated to use the new `fmae` function right?

GabrielMajeri commented:
@gnzlbg I was going to wait for #123 to land so I can use `mul_adde` .

I'm not planning to get this merged however, not until we manage to beat ISPC in `aobench` at least.

GabrielMajeri commented:
I'd decided to compare oranges to oranges and use `libm` for both ISPC and Rust, to avoid comparing using `libsleef` vs. ISPC's inlined math functions.

The results are in:
```
cargo run --release 1920 1080 --algo tiled
time: 4170 ms
```
```
cargo run --release 1920 1080 --algo ispc
time: 4532 ms
```
So Rust is faster than ISPC. I don't know if it's due to a lack of inlining, or some issue with sleef, but it's clear there's not much more we can do to optimize the Rust code itself.

gnzlbg commented on examples/aobench/src/vector.rs:
```diff
@@ -81,7 +81,7 @@ cfg_if! {
-                if is_x86_feature_detected!("avx2") {
+                if is_x86_feature_detected!("avx2") && is_x86_feature_detected!("fma") {
                     ao_avx2(scene, nsubsamples, img);
                 } else if is_x86_feature_detected!("avx") {
```
I think we should also have a branch here for AVX+FMA (and also one for AVX without FMA)

gnzlbg commented on examples/stencil/src/lib.rs:
```diff
@@ -131,22 +133,24 @@ fn assert_data_eq(a: &Data, b: &Data) {
 
-                assert_eq!(
-                    a.vsq[idx], b.vsq[idx],
+                const EPSILON: f32 = 1E-4;
```
Is this the smallest absolute epsilon that works?

GabrielMajeri replied:
Yes, at least on my computer.

gnzlbg commented on examples/stencil/src/simd.rs:
```diff
@@ -158,7 +170,8 @@ pub fn x8(
+        if is_x86_feature_detected!("avx2") && is_x86_feature_detected!("fma")
+        {
             #[rustfmt::skip]
             x8_impl_avx2(t0, t1, x0, x1, y0, y1, z0, z1, n_x, n_y, n_z,
```
Same here for avx +- fma

gnzlbg commented:
Note that ISPC inline math functions are not as precise as sleef, libm, SVML, etc. 

Let me know when you finish the nitpicks and this is green so that we can merge this. I think it looks really good. I'll run the benchmarks on some systems and update the readme's afterwards (you can also add the results for your own systems in a follow up PR if you want).

gnzlbg commented:
Also: THANKS for doing this!

GabrielMajeri commented:
Alright, just fixed a _little_ **big** issue with the RNG: those functions were _not_ marked as `#[inline(always)]`, so the `target_feature` attribute would end up not applying to them, so we would end up with AVX2 code calling into a function built with SSE2 instructions.

<hr/>

I think I've addressed most of the issues I could, some final bechmark results here (with `libm`, in order from slowest to fastest):
```
ispc: 4542 ms
vector: 4355 ms
tiled: 4117 ms

ispc tasks: 1227 ms
vector_par:  1233 ms
tiled_par: 1170 ms
```

<hr/>

Side note: would it be fair to modify `build.rs` to force ISPC to always use the system's math library? After all, this repository is not about benchmarking's ISPC fast math library

GabrielMajeri commented:
@gnzlbg Clippy failed, but I don't think it's related to this PR.

> I'd prefer if you do this in a different PR, because it applies to many benchmarks and I'd like to merge this ;)

Alright, I'll see if I can't get cross-language inlining to work, maybe we can use sleef to beat ISPC's math library.

gnzlbg commented:
> Clippy failed, but I don't think it's related to this PR.

I'll fix that :) 

> Alright, I'll see if I can't get cross-language inlining to work, maybe we can use sleef to beat ISPC's math library.

I have a branch where I started doing this the other day. I can upload it if you want to pick it up from there.

gnzlbg commented:
@GabrielMajeri i'm gnzlbg on IRC
